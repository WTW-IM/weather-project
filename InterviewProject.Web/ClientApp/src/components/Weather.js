import React from "react";
import Autocomplete from "react-autocomplete";
import { WeatherApiInstance } from "../service/WeatherApiService";
import useDebounce from "./debounce/DebounceHook";
import { Row, Col } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import { AuthServiceInstance } from '../service/AuthService';
import { AxiosLoadingRenderProps } from './axios/Wrapper';

export function Weather(props) {
  const [query, saveQuery] = React.useState("");
  const [locations, saveLocations] = React.useState([]);
  const [forecasts, saveForecasts] = React.useState([]);
  const [isAuthorized, saveIsAuthorized] = React.useState(false);

  React.useEffect(() => {
   
    if (!isAuthorized) {
      const checkAuth = async () => {
        AuthServiceInstance.registerLoginCallback(user => {
          if (user && user.access_token) {
            saveIsAuthorized(true);
          }
        });
        await AuthServiceInstance.authGuard();
      };
      checkAuth();
    }

  }, [isAuthorized]);

  const formatTemp = temp => {
    return Math.round(temp);
  };

  const convertToFahrenheit = temp => {
    return temp * 1.8 + 32;
  };

  const renderForecasts = () => {
    let content;
    if (forecasts && forecasts.length > 0) {
      content = (
        <table className="table table-striped" aria-labelledby="tabelLabel">
          <thead>
            <tr>
              <th>Date</th>
              <th>Temp. (C)</th>
              <th>Temp. (F)</th>
              <th>Summary</th>
            </tr>
          </thead>
          <tbody>
            {forecasts.map(forecast => (
              <tr key={forecast.applicable_date}>
                <td>{forecast.displayDate}</td>
                <td>{formatTemp(forecast.the_temp)}</td>
                <td>{formatTemp(convertToFahrenheit(forecast.the_temp))}</td>
                <td>{forecast.weather_state_name}</td>
              </tr>
            ))}
          </tbody>
        </table>
      );
    }
    return content;
  };

  const clearSearch = e => {
    e.preventDefault();
    saveQuery("");
    saveLocations([]);
    saveForecasts([]);
  };

  const debouncedSearch = useDebounce(query, 20);

  return (
    <div>
      <AxiosLoadingRenderProps/>
      <Row>
        <Col>
          <h1 id="tabelLabel">Weather Forecast</h1>
          <p>This component fetches weather data from metaweather.com</p>
        </Col>
      </Row>
      <Row>
        <Col>
          <label htmlFor="locations-autocomplete">Search for a city:</label>
        </Col>
        <Col>
          <Autocomplete
            className="w-100"
            inputProps={{
              id: "locations-autocomplete",
              style: { width: "100%" }
            }}
            wrapperStyle={{
              position: "relative",
              display: "inline-block",
              width: "100%"
            }}
            value={query}
            items={locations}
            getItemValue={item => item.title}
            onSelect={async (value, item) => {
              const forecast = await WeatherApiInstance.getForecast(item.woeid);
              saveForecasts(forecast.consolidated_weather);
              saveQuery(value);
              saveLocations([item]);
            }}
            onChange={async (event, value) => {
              saveQuery(value);
              saveLocations([]);
              saveForecasts([]);
              if (debouncedSearch && debouncedSearch.length > 1) {
                const items = await WeatherApiInstance.locationSearch(value);
                if (items) {
                  saveLocations(items);
                }
              }
            }}
            renderMenu={children => <div className="menu">{children}</div>}
            renderItem={(item, isHighlighted) => {
              return (
                <div
                  style={{ background: isHighlighted ? "lightgray" : "white" }}
                >
                  {item.title}
                </div>
              );
            }}
          />
        </Col>
        <Col>
          <a className="float-left" onClick={clearSearch}>
            <FontAwesomeIcon icon={faTimes} />
          </a>
        </Col>
      </Row>
      <Row className="pt-5">
        <Col>{renderForecasts()}</Col>
      </Row>
    </div>
  );
}