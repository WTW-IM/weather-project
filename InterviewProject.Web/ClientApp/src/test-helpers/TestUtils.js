import { act } from 'react-dom/test-utils';

//work around for enzyme issue:
//https://github.com/airbnb/enzyme/issues/2073
export const waitForComponentToPaint = async (wrapper) => {
  act(async () => {
    await new Promise(resolve => setTimeout(resolve, 0));
    wrapper.update();
  });
};

//work around for fontAwesome snapshot issue
//https://github.com/FortAwesome/react-fontawesome/issues/194
export const fontAwesomeTitleMock = () => {
  const mockMath = Object.create(global.Math);
  mockMath.random = () => 0.5;
  global.Math = mockMath;
};
