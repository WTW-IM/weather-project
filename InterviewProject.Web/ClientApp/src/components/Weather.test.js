import * as React from 'react';
import { Weather } from './Weather';
import { waitForComponentToPaint, fontAwesomeTitleMock } from '../test-helpers/TestUtils';
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter() });

describe("Weather Component", () => {
    let wrapper;
    
    beforeEach(() => {
        fontAwesomeTitleMock();
        wrapper = Enzyme.mount(
            <Weather/>);
       // waitForComponentToPaint(wrapper);
    });

    it("it should match the snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
   

});
