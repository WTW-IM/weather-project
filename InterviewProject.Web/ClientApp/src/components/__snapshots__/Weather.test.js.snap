// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Component it should match the snapshot 1`] = `ReactWrapper {}`;
